### PR TITLE
yarpserver class created + yarpserver_main updated

### PR DESCRIPTION
**Pull request by aerydna**
`YarpServerThread` class created in order to use yarpserver as a standard executable or in a plugin (example: yarpserverfixture)

### CODE REVIEW

File: src/libYARP_serversql/src/yarpserver.cpp [hunk 1]
```diff
@@ -291,11 +273,58 @@ yarpserversql_API int yarpserver_main(int argc, char *argv[]) {
...
+
+    Network          yarp(yarp::os::YARP_CLOCK_SYSTEM);
+    YarpServerThread yServer;
+
+    std::signal(SIGINT, [](int sig){should_exit = true;});
+    yServer.configure(argc, argv);
+    yServer.start();
+    while(!should_exit)
+    {
+
```
Comment (randaz81) [inline src/libYARP_serversql/src/yarpserver.cpp]: busy loop. Insert a delay.

File: src/libYARP_serversql/src/yarpserver.cpp [hunk 2]
```diff
@@ -174,41 +176,32 @@ class NameServerContainer : public ComposedNameService {
...
-    if (b.get(1).asString()!="10.0.0.10") {
-        fprintf(stderr, "Sorry, please update YARP version");
-        ::exit(1);
-    }
 
-    Property options;
+    Property            options;
+    bool                silent(false);
+    FILE*               out;
+    NameServerContainer nc; if (!nc.open(options)) return;
```
Comment (Nicogene) [inline src/libYARP_serversql/src/yarpserver.cpp]: I would move this line after the next if-else block, to change as little as possible.

File: src/libYARP_serversql/src/yarpserver.cpp [hunk 3]
```diff
@@ -174,41 +176,32 @@ class NameServerContainer : public ComposedNameService {
...
+
     options.fromCommand(argc, argv, false);
+
     silent = options.check("silent");
+    out    = silent ? tmpfile() : stdout;
+
+    nc.setSilent(silent);
```
Comment (Nicogene) [inline src/libYARP_serversql/src/yarpserver.cpp]: Same as above.

**Comment by drdanz** [comment]
Please test and confirm that this works on windows as well

**Comment by traversaro** [comment]
Just to understand, what problem is solved by this PR w.r.t. to the existing yarpserver fixture plugin?

File: src/libYARP_serversql/include/yarp/serversql/yarpserversql.h [hunk 4]
```diff
@@ -28,6 +29,23 @@
...
 #endif
 
+class YarpServerThread : public yarp::os::Thread
+{
+public:
+    virtual ~YarpServerThread(){}
+    void configure(int inArgc, char** inArgv)
+    {
+        argc = inArgc;
+        argv = inArgv;
```
Comment (traversaro) [inline src/libYARP_serversql/include/yarp/serversql/yarpserversql.h]: If this is intended to be a public API, we should clarify that the array of strings pointed by `inArgv` should be valid for all the life of the `YarpServerThread`. This is obvious if `argc` and `argv` are the arguments of the `main`, but in general it is not true, so we should it at least mention this in the documentation (if this is in a public API).

**Comment by traversaro** [comment]
With this PR, the `yarpserver` executable is unable to process command line arguments, try for example `yarpserver --help`.

**Comment by aerydna** [comment]
@traversaro `yarpserver --help` works fine on both windows and linux. can it be macOS problem only? still you confirm it to not working?

**Comment by aerydna** [comment]
ok.. the server was not closing correctly only with the `--help` parameter

File: src/libYARP_serversql/src/yarpserver.cpp [hunk 5]
```diff
@@ -174,41 +176,28 @@ class NameServerContainer : public ComposedNameService {
...
-    }
-    nc->goPublic();
-    return nc;
-}
-
-yarpserversql_API int yarpserver_main(int argc, char *argv[]) {
+void YarpServerThread::run()
+{
     // check if YARP version is sufficiently up to date - there was
     // an important bug fix
```
Comment (traversaro) [inline src/libYARP_serversql/src/yarpserver.cpp]: I think this check was removed, so we should remove also this comment.

File: src/rtf-plugins/fixture-managers/yarpserver/yarpserverfixture.cpp [hunk 6]
```diff
@@ -12,34 +12,19 @@
...
-};
+//using namespace yarp::name;
 
 class YarpNameServer : public FixtureManager
 {
 public:
     yarp::os::Network net;
     YarpNameServer() {};
-    virtual ~YarpNameServer() {};
+    virtual ~YarpNameServer() {yServer.stop();};
```
Comment (traversaro) [inline src/rtf-plugins/fixture-managers/yarpserver/yarpserverfixture.cpp]: Why don't we keep this blocking stop in the teardown?

File: src/rtf-plugins/fixture-managers/yarpserver/yarpserverfixture.cpp [hunk 7]
```diff
@@ -12,34 +12,19 @@
 #include <rtf/dll/Plugin.h>
 #include <yarp/serversql/yarpserversql.h>
 #include <yarp/os/LogStream.h>
+#include <yarp/os/impl/NameClient.h>
+#include <cmath>
```
Comment (traversaro) [inline src/rtf-plugins/fixture-managers/yarpserver/yarpserverfixture.cpp]: Why `<cmath>` is included?

File: src/libYARP_serversql/src/yarpserver.cpp [hunk 8]
```diff
@@ -224,62 +213,56 @@ yarpserversql_API int yarpserver_main(int argc, char *argv[]) {
         printf("  --no-web-cache           Reload pages from file for each request.\n");
         printf("  --ros                    Delegate pub/sub to ROS name server.\n");
         printf("  --silent                 Start in silent mode.\n");
-        return 0;
-    } else {
-        fprintf(out, "Call with --help for information on available options\n");
+        this->stop();
```
Comment (traversaro) [inline src/libYARP_serversql/src/yarpserver.cpp]: The `stop` method of the Thread should not be called by the thread itself, see https://github.com/robotology/yarp/blob/e0ca8e6f73c37286734d06e3afb8015529ee36b9/src/libYARP_OS/include/yarp/os/Thread.h#L76 . Returning from the run should be sufficient.

File: src/libYARP_serversql/src/yarpserver.cpp [hunk 9]
```diff
@@ -291,11 +274,58 @@ yarpserversql_API int yarpserver_main(int argc, char *argv[]) {
...
+
+    std::signal(SIGINT, [](int sig){should_exit = true;});
+    yServer.configure(argc, argv);
+    yServer.start();
+    while(!should_exit && !yServer.isStopping())
+    {
+        yarp::os::Time::delay(0.1);
+    }
+    yServer.stop();
     return 0;
```
Comment (traversaro) [inline src/libYARP_serversql/src/yarpserver.cpp]: If I understand the code correctly, there is a regression, as the yarpserver will now always return 0 regardless if it started successfully or not. See https://github.com/robotology/yarp/issues/395 .

**Comment by traversaro** [comment]
`yarpserver --help` is now correctly working on my setup.

**Comment by traversaro** [comment]
It seems to me that with this PR there is a sensible delay (~1 seconds) when closing yarpserver from the terminal.

File: src/rtf-plugins/fixture-managers/yarpserver/yarpserverfixture.cpp (hunk 6 above)
Comment (drdanz) [reply src/rtf-plugins/fixture-managers/yarpserver/yarpserverfixture.cpp]: +1

**Comment by drdanz** [comment]
To be honest, I don't like this patch, this adds an extra layer of complication over something that is already over complicated.
The extra thread is completely unneeded for the normal setup, it is only required for the fixture, I'd rather find another way to deal with this...

**Comment by drdanz** [comment]
btw, this FixtureManager should definitely implement "bool check()" calling `checkNetwork()`

File: src/rtf-plugins/fixture-managers/yarpserver/yarpserverfixture.cpp (hunk 6 above)
Comment (aerydna) [reply src/rtf-plugins/fixture-managers/yarpserver/yarpserverfixture.cpp]: for what i remember it is better this to be the very last step. rtf calls all the teardowns and then it destroy all the objects.. so if we want the yarpserver shutting down to to be alive during all the other fixture disruption it it better to leave this as it is.

Comment (traversaro) [reply src/rtf-plugins/fixture-managers/yarpserver/yarpserverfixture.cpp]: If the order of teardown of the fixtures in RTF is reversed w.r.t. to the order of setup, this should not be a problem, as all the fixture using yarp need to be setup after the yarpserver, and so they will be teardown before the yarpserver. This should be the case given https://github.com/robotology/robot-testing/pull/69/files .

**Comment by aerydna** [comment]
the patch has been under-complicatized by removing the thread inheritance from the yarp server class and moving it to the fixture. the side effect is that if in the future someone want to use it in a separate thread(the yarpserver class), will have to basically duplicate the yarp server fixture code the "thread" it.

File: src/rtf-plugins/fixture-managers/yarpserver/yarpserverfixture.cpp [hunk 10]
```diff
@@ -49,7 +49,13 @@ class YarpNameServer : public FixtureManager
     {
 
         yServer.configure(argc, argv);
-        return yServer.start();
+        yServer.start();
+        double time = yarp::os::Time::now();
```
Comment (traversaro) [inline src/rtf-plugins/fixture-managers/yarpserver/yarpserverfixture.cpp]: I suggest to always use the system clock here:
~~~
double time = yarp::os::SystemClock::nowSystem()
~~~

File: src/rtf-plugins/fixture-managers/yarpserver/yarpserverfixture.cpp [hunk 11]
```diff
@@ -49,7 +49,13 @@ class YarpNameServer : public FixtureManager
...
+        while(!net.checkNetwork() || (yarp::os::Time::now() - time) < YSERVFIXT_TIMEOUT)
+        {
+            yarp::os::Time::delay(0.1);
```
Comment (traversaro) [inline src/rtf-plugins/fixture-managers/yarpserver/yarpserverfixture.cpp]: Here as well: 
~~~
yarp::os::SystemClock::delaySystem(0.1)
~~~
Furthermore, I would suggest to choose a smaller waiting time, such as `0.001`

**Pull request merged**